File: SecGest/proyecto frontend/src/app/service/task-category-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Category, CategoryRequest } from '../model/category';

@Injectable({
  providedIn: 'root',
})
export class TaskCategoryService {
  // Microservicio "categories" (puerto 8083), independiente de "users" (8081) y "tasks" (8082).
  private categoriesUrl = 'http://localhost:8083/api/categories';

  constructor(private http: HttpClient) {}

  // Publico: solo categorias activas (catalogo de cara al cliente).
  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(this.categoriesUrl);
  }

  // Solo admin: activas e inactivas, para poder gestionarlas todas.
  getAllCategoriesForAdmin(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.categoriesUrl}/all`);
  }

  createCategory(req: CategoryRequest): Observable<Category> {
    return this.http.post<Category>(this.categoriesUrl, req);
  }

  updateCategory(id: number, req: CategoryRequest): Observable<Category> {
    return this.http.put<Category>(`${this.categoriesUrl}/${id}`, req);
  }

  // Baja logica: la categoria deja de aparecer en el catalogo publico, pero
  // las tareas que ya la referencian siguen pudiendo mostrar su nombre.
  deactivateCategory(id: number): Observable<Category> {
    return this.http.delete<Category>(`${this.categoriesUrl}/${id}`);
  }

  activateCategory(id: number): Observable<Category> {
    return this.http.put<Category>(`${this.categoriesUrl}/${id}/activate`, {});
  }
}

File: SecGest/categories/src/main/java/com/igarciamen/categories/service/TaskCategoryService.java
package com.igarciamen.categories.service;

import com.igarciamen.categories.model.TaskCategory;
import com.igarciamen.categories.payloads.request.CategoryRequest;
import com.igarciamen.categories.repository.TaskCategoryRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Service
public class TaskCategoryService {

    private final TaskCategoryRepository categoryRepo;

    public TaskCategoryService(TaskCategoryRepository categoryRepo) {
        this.categoryRepo = categoryRepo;
    }

    // Catalogo publico: solo categorias activas.
    public List<TaskCategory> listPublic() {
        return categoryRepo.findAllByActiveTrueOrderByNameAsc();
    }

    // Panel del admin: todas, activas e inactivas.
    public List<TaskCategory> listAllForAdmin() {
        return categoryRepo.findAllByOrderByNameAsc();
    }

    public TaskCategory getById(Long id) {
        return categoryRepo.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Category not found: " + id));
    }

    public TaskCategory create(CategoryRequest req) {
        if (categoryRepo.existsByNameIgnoreCase(req.getName())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "Category already exists: " + req.getName());
        }
        TaskCategory category = new TaskCategory(req.getName(), req.getDescription());
        applyOptionalFields(category, req);
        category.setActive(true);
        return categoryRepo.save(category);
    }

    public TaskCategory update(Long id, CategoryRequest req) {
        TaskCategory category = getById(id);
        category.setName(req.getName());
        category.setDescription(req.getDescription());
        applyOptionalFields(category, req);
        return categoryRepo.save(category);
    }

    // Baja logica: no se borra la fila, para que las tareas que ya referencian
    // esta categoria (por id) sigan pudiendo resolver su nombre via CategoryClient.
    public TaskCategory deactivate(Long id) {
        TaskCategory category = getById(id);
        category.setActive(false);
        return categoryRepo.save(category);
    }

    public TaskCategory activate(Long id) {
        TaskCategory category = getById(id);
        category.setActive(true);
        return categoryRepo.save(category);
    }

    private void applyOptionalFields(TaskCategory category, CategoryRequest req) {
        category.setIcon(req.getIcon());
        category.setColorHex(req.getColorHex());
        category.setImageUrl(req.getImageUrl());
        category.setBasePrice(req.getBasePrice());
        category.setEstimatedMinutes(req.getEstimatedMinutes());
    }
}
